### PR TITLE
Fix memory leak in enumerate iterator

### LINKED ISSUE #3473: Memory leak when using enumerate in a jit function
## Reporting a bug

- [X] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [X] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

There seem to be a memory leak caused by the use of `enumerate` in a `jit` function.

The following script (requires the `psutil` package.) prints the memory usage at each iteration and it's clearly increasing.

```python
import psutil
from numba import (jit, jitclass, uint32)

@jitclass([
    ('attr', uint32),
])
class JitClass:
    def __init__(self):
        pass


@jit  # same with njit
def f():
    cs = [JitClass() for i in range(1000)]

    # leak
    for i, c in enumerate(cs):
        c.attr = 1234

    # No enumerate, no leak
    # for c in cs:
    #     c.attr = 1234

    
p = psutil.Process()
for _ in range(10000):
    f()
    # leak proportional to the size of cs
    print(f"memory used by program: {p.memory_info().rss / 1e6} MB")
```

This seems to be coupled with the list comprehension because the following doesn't leak:

```
@jit
def f():
    c = JitClass()
    for i, j in enumerate(range(10)):
        c.attr = 1234
```

Not sure if this is related but using list comprehensions has some weird side effect (I also opened #3472)

### PR DESCRIPTION
**Pull request by sklam**
Fix #3473.
Based on #3672.

* Add missing decref into enumerate iterator `next`
* Add missing zero-fill.  When `alloca_once(..., zfill=True)` is used in loop, the expectation is that the data is zero-filled at the use-site.  But, the zero-filling is located in the entry block only, causing the state from the previous iteration to leak into the future iteration.

### CODE REVIEW

File: numba/targets/iterators.py [hunk 1]
```diff
@@ -61,6 +61,8 @@ def iternext_enumerate(context, builder, sig, args, result):
 
     with builder.if_then(is_valid):
         srcval = srcres.yielded_value()
+        if context.enable_nrt:
```
Comment (stuartarchibald) [inline numba/targets/iterators.py]: Want to add a comment here for future visitors of this code path?

File: numba/targets/tupleobj.py [hunk 2]
```diff
@@ -188,6 +188,8 @@ def iternext_unituple(context, builder, sig, args, result):
                                        types.intp)
         getitem_out = getitem_unituple(context, builder, getitem_sig,
                                        [tup, idx])
+        if context.enable_nrt:
```
Comment (stuartarchibald) [inline numba/targets/tupleobj.py]: Want to add a comment here for future visitors of this code path?

File: numba/targets/iterators.py (hunk 1 above)
Comment (sklam) [reply numba/targets/iterators.py]: done

File: numba/targets/tupleobj.py (hunk 2 above)
Comment (sklam) [reply numba/targets/tupleobj.py]: done

**Pull request merged**